f/sliceLazy.js

```diff
@@ -0,0 +1,36 @@
+/**
+ *
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+function* slice(arr, start, end) {
+  if (typeof start !== 'number') start = 0;
+  if (start < 0) start = arr.length + start;
+  if (start < 0) start = 0;
+  if (start >= arr.length) return;
+
+  if (typeof end !== 'number') end = arr.length;
```

Comment:
Sorry, I don't know much about generators.

Well, I was trying to duplicate the behavior of Array.prototype.slice. But, if is not possible to know the length of the array, I think that the generator with a ring buffer will not return the same result.

For example:

```javascript
[1, 2, 3, 4, 5].slice(5, undefined); // start >= length returns []
[1, 2, 3, 4, 5].slice(undefined, -5); // -end >= length returns []

slice([1, 2, 3, 4, 5], 5, undefined); // Will return [1, 2, 3, 4, 5]
slice([1, 2, 3, 4, 5], undefined, -5); // Will return [1, 2, 3, 4, 5]
```

Right?

Also, I forgot to add it to the tests, but start parameter can be negative too:

```javascript
[1, 2, 3, 4, 5].slice(-5, undefined); // -start >= length returns [1, 2, 3, 4, 5] (full array)
[1, 2, 3, 4, 5].slice(-3, undefined); // returns [3, 4, 5]
```

And, I noticed that Array.prototype.slice allows any type of value as index.
I made some more tests and it currently only fails if an object is passed as index.
https://gist.github.com/RockoDev/1a4eb2081691dd7096e06601605d28dc

This function it's a bit more complex than I thought, hehe.